pkg/yurthub/certificate/pki.go

```diff
@@ -23,14 +23,76 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"time"
 
 	"k8s.io/client-go/util/certificate"
+	"k8s.io/klog/v2"
+
+	"github.com/openyurtio/openyurt/pkg/projectinfo"
 )
 
-// GenTGenTLSConfigUseCertMgrAndCertPool generates a TLS configuration
+// hubServerCertMgr is the certificate manager for server usage, which will cache the old hub-server-certificate,
+type hubServerCertMgr struct {
+	// the path of old hub server certificate
+	oldCertFile string
+	// the cached old hub server certificate
+	oldCert *tls.Certificate
+	// the actual hub certMgr
+	hubCertMgr certificate.Manager
+}
+
+// newHubServerCertMgr creates a certificate manager for the yurthub-server
+func newHubServerCertMgr(certDir string, hubCertMgr certificate.Manager) hubServerCertMgr {
+	var oldCertFile string
+	var oldCert *tls.Certificate
+
+	store, err := certificate.NewFileStore(fmt.Sprintf("%s-server", projectinfo.GetHubName()), certDir, certDir, "", "")
+	if err != nil {
+		oldCertFile = ""
+		oldCert = nil
+	} else {
+		oldCertFile = store.CurrentPath()
+		oldCert, err = store.Current()
+		if err != nil {
+			oldCert = nil
+		}
+	}
+
+	return hubServerCertMgr{
+		oldCertFile: oldCertFile,
+		oldCert:     oldCert,
+		hubCertMgr:  hubCertMgr,
+	}
+}
+
+// Current is used to obtain the certificate for hub server usage.
+// If the current certificate of hub certMgr is invalid (if the key usages of certificate doesn't contain the UsageServerAuth),
+// the old certificate in the yurthub-server-current.pem will be used.
+func (scm *hubServerCertMgr) Current() *tls.Certificate {
+	cert := scm.hubCertMgr.Current()
+	if cert == nil {
+		return &tls.Certificate{Certificate: nil}
+	}
+
+	if hasServerAuthUsage(cert) {
+		return cert
+	}
+
+	klog.V(3).Infof("the %s certificate is not valid for server, try to use the old server certificate in %s",
+		projectinfo.GetHubName(), scm.oldCertFile)
+	if scm.oldCert != nil && time.Now().Before(scm.oldCert.Leaf.NotAfter) {
+		return scm.oldCert
+	} else {
+		klog.V(3).Infof("failed to get the old %s server certificate, wait for the new hub certificate", projectinfo.GetHubName())
+	}
+	return &tls.Certificate{Certificate: nil}
+}
+
+// GenTLSConfigUseCertMgrAndCertPool generates a TLS configuration
 // using the given certificate manager and x509 CertPool
 func GenTLSConfigUseCertMgrAndCertPool(
```

Comment:
how about delete `GenUseCertMgrAndTLSConfig` func and remove routine into `GenTLSConfigUseCertMgrAndCertPool `

Reply:
The code is updated, please take a review.